File: code_snippets/java/src/main/java/develop/Greeter.java
package develop;

import dev.restate.sdk.*;
import dev.restate.sdk.annotation.*;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.http.vertx.RestateHttpServer;

@VirtualObject
public class Greeter {

  private static final StateKey<Integer> COUNT = StateKey.of("count", Integer.class);

  @Handler
  public String greet(ObjectContext ctx, String greeting) {
    // Get the count and increment it
    int count = ctx.get(COUNT).orElse(1);
    ctx.set(COUNT, count + 1);

    // Send the response back
    return greeting + " " + ctx.key() + ", for the " + count + " time!";
  }

  public static void main(String[] args) {
    // Start the Restate Endpoint HTTP Server
    RestateHttpServer.listen(Endpoint.bind(new Greeter()));
  }
}

File: code_snippets/kotlin/src/main/kotlin/develop/Greeter.kt
package develop

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.VirtualObject
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.kotlin.endpoint.*

@VirtualObject
class Greeter {

  companion object {
    private val COUNT = stateKey<Int>("count")
  }

  @Handler
  suspend fun greet(ctx: ObjectContext, greeting: String): String {
    // Get the count and increment it
    val count = ctx.get(COUNT) ?: 1
    ctx.set(COUNT, count + 1)

    // Send the response back
    return "$greeting ${ctx.key()} for the $count time!"
  }
}

fun main() {
  RestateHttpServer.listen(endpoint { bind(Greeter()) })
}
